src/libmeasurement_kit/nettests/base_test.cpp

```diff
@@ -98,25 +99,59 @@ BaseTest &BaseTest::on_destroy(Delegate<> cb) {
     return *this;
 }
 
+static void start_internal_(Var<Runnable> &&r, std::promise<void> *promise,
+                            Callback<> &&callback) {
+    // Note:
+    //
+    // 1. we wait in the default tasks queue for our turn: this guarantees
+    //    that "big" operations like tests and tasks are sequenced.
+    //
+    // 2. we move all context inside the first lambda callback so that the
+    //    thread that will run I/O has exclusive ownership.
+    //
+    // 3. run_with_initial_event() is blocking: this means that the lifecycle
+    //    of `r` is guaranteed to be as long as the one of the reactor.
+    //
+    // 4. we _explicitly_ leave the reactor loop on success, but, even in
+    //    case the final event is not reached because of a bug, the reactor
+    //    is anyway going to exit the loop when it is out of events.
+    //
+    // 5. the `promise`, if present, allows to make the test synchronous,
+    //    while the callback allows to make it asynchronous.
+    assert(!r->reactor);
+    Worker::default_tasks_queue()->run_in_background_thread(
+          [ r = std::move(r), promise, callback = std::move(callback) ] {
+              r->reactor = Reactor::make();
+              r->reactor->run_with_initial_event([&]() {
+                  r->begin([&](Error) {
+                      r->end([&](Error) {
+                          r->reactor->stop();
+                          if (callback) {
+                              callback();
+                          }
+                      });
+                  });
+              });
+              if (promise != nullptr) {
+                  promise->set_value();
+              }
+          });
+}
+
 void BaseTest::run() {
-    // Note: here we MUST point to a fresh reactor which we know for sure is
-    // not already being used otherwise we cannot run the test
-    assert(not runnable->reactor);
-    Var<Reactor> reactor = Reactor::make();
-    runnable->reactor = reactor;
-    reactor->loop_with_initial_event([&]() {
-        runnable->begin([&](Error) {
-            runnable->end([&](Error) {
-                reactor->call_soon([&]() {
-                    reactor->stop();
-                });
-            });
-        });
-    });
+    std::promise<void> promise;
+    std::future<void> future = promise.get_future();
+    // Note: using `std::move` to invalidate the `runnable` such that it is
+    // not possible to start another test from the same object.
```

Comment:
Before that was not done explicitly, but I always meant for that to be the case.